group_project_v2/mixins.py

```diff
@@ -125,7 +126,201 @@ def real_user_id(self, anonymous_student_id):
+        allowed_roles = set(self.ta_roles)
+        return bool(allowed_roles & granted_roles)
+
+    def check_ta_access(self, user_id, course_id):
+        """
+        :return: None
+        :raise GroupworkAccessDeniedError: If user can't access a groupwork item
+        if he is not part of the team.
+        """
+        if not self.is_user_ta(user_id, course_id):
+            raise GroupworkAccessDeniedError("User can't access this group work")
+
+    def get_organization_filter_for_user(self, user_id, additional_filter=None):
+        """
+        :param Iterable[int] additional_filter:
+            Iterable of organization ids that are filtered additionally.
+        :return: Returns object that can be used if user_id has access to
+                 an organization. This callable returns bool an expects to be
+                 called with organization id.
+        :rtype: Callable[[int], bool]
+        """
+        if self._can_user_access_all_orgs(user_id):
+            allowed_org_ids = None
+        else:
+            allowed_orgs = self.project_api.get_user_organizations(user_id)
+            allowed_org_ids = set(org['id'] for org in allowed_orgs)
+        return self.OrganizationFilter(self.project_api, user_id, allowed_org_ids, additional_filter)
+
+    class OrganizationFilter(object):
+
+        """
+        A class that checks whether an user can access an organization.
+
+        Contains two sets: one with organizations user has access to, and second
+        with organizations that user has explicitily filtered for.
+
+        Value None means "all organizations" --- since there can be a large
+        number of organizations, we avoid keeping "all organizations" that way.
+        """
+
+        def __init__(self, project_api, user_id, allowed_org_ids, filter_org_ids=None):
+            """
+            :param TypedProjectApi project_api:
+            :param user_id:
+            :param set[int] allowed_org_ids:
+            :param set[int] filter_org_ids:
+            """
+            self.project_api = project_api
+            self.user_id = user_id
+            self.filter_org_ids = set(filter_org_ids) if filter_org_ids is not None else None
+            """
+            :type: set[int] or None
+            """  # pylint: disable=pointless-string-statement
+            self.allowed_org_ids = set(allowed_org_ids) if allowed_org_ids is not None else None
+            """
+            :type: set[int] or None
+            """  # pylint: disable=pointless-string-statement
+
+        def can_access_other_user(self, user_id):
+            """
+            :param user_id:
+            :return: True if this user can access user with id of ``user_id``.
+                     This user is the user for which this ``OrganizationFilter``
+                     was created (see: self.user_id).
+            :rtype bool:
+            """
+            orgs = self.project_api.get_user_organizations(user_id)
```

Comment:
@e-kolpakov I have added the memoization, if this still turns out to be a performance problem we could query users of every organisation current user has access to (unless he has access to all orgs), I mean ``users`` property of this endpoint:  ``http://apros.mcka.local/api/server/organizations/{user_id}/``. Which should be faster unless we happen to have admin that administers hundreds of organisations each couple of thousand users strong (this might be unlikelu).